spec/models/user_spec.rb

```diff
@@ -0,0 +1,78 @@
+require "rails_helper"
+
+RSpec.describe User, type: :model do
+
+  it "has a valid user" do
+    expect(build(:user)).to be_valid
+  end
+
+  describe "associations" do
+
+    it "has many bookings" do
+      is_expected.to have_many(:bookings).dependent :destroy
+    end
+
+    it "has many likes" do
+      is_expected.to have_many(:likes).dependent :destroy
+    end
+
+    it "has many reviews" do
+      is_expected.to have_many(:reviews).dependent :destroy
+    end
+
+    it "has many comments" do
+      is_expected.to have_many(:comments).dependent :destroy
+    end
+
+    context "presence" do
+      it { should validate_presence_of :name }
+      it { should validate_presence_of :phonenumber }
+    end
+  end
+
+  describe "roles" do
+    it "are user or admin" do
+      is_expected.to define_enum_for(:role).with_values %i(user admin)
+    end
+  end
+
+  describe "validate" do
+    let(:user){FactoryBot.create :user}
```

Comment:
code `let` này để làm gì em?

phân biệt `let` và `let!` nào

Reply:
let (lazy-evaluated): chỉ chạy khi phương thức mà nó định nghĩa được gọi.
let!: thì chạy trước ngay trước cả khi phương thức mà nó định nghĩa được gọi.
let!: khi cần tính toán kiểu như: count += 1.